python/cudf_polars/cudf_polars/experimental/groupby.py

```diff
@@ -211,9 +211,13 @@ def _(
     # Check if we are dealing with any high-cardinality columns
     post_aggregation_count = 1  # Default tree reduction
     groupby_key_columns = [ne.name for ne in ir.keys]
+
+    # TODO: determine if InMemoryExecutor can reach here
+    assert ir.config_options.executor.name == "streaming"
```

Comment:
@rjzamora I confused myself between "executors" and "schedulers", and which are possible to have where.

Is it ever possible for the `in-memory` executor to reach this code (in `experimental/groupby.py`)?

This matters because `ir.config_options.executor` has a union type: `InMemoryExecutor | StreamingExecutor`. At the moment, I haven't defined `InMemoryExecutor.cardinality_factory`, because I'd like to avoid defining configuration options that don't *actually* have an effect.

Similar question for the 3 other places where I've left "TODO: determine if InMemoryExecutor".

Reply:
> Is it ever possible for the in-memory executor to reach this code (in experimental/groupby.py)?

No, that would be a bug. You can think of the `cudf_polars.experimental` module as being the `cudf_polars.streaming` module - This directory only contains code that is used by the streaming/experimental executor.

Reply:
Great, thanks for confirming.

In that case, I think I'll just keep the `asserts` like I have have here. Catching this a type-checking time would be possible if we made `IR` generic over the type of `config_options.executor`. IMO that's not worth it at the moment.